File: backend/app/api/evidence.py
"""
Evidence API endpoints
POST /evidence/presigned-url - Generate S3 presigned upload URL
POST /evidence/upload-complete - Notify upload completion and create evidence record
GET /evidence/{evidence_id} - Get evidence detail with AI analysis
GET /evidence/{evidence_id}/status - Get evidence processing status
POST /evidence/{evidence_id}/retry - Retry failed evidence processing

Note: GET /cases/{case_id}/evidence is in cases.py (follows REST resource nesting)
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional

from app.db.session import get_db
from app.db.schemas import (
    PresignedUrlRequest,
    PresignedUrlResponse,
    UploadCompleteRequest,
    UploadCompleteResponse,
    EvidenceDetail
)
from app.services.evidence_service import EvidenceService
from app.core.dependencies import get_current_user_id
from app.middleware import NotFoundError, PermissionError


# Response models for new endpoints
class EvidenceStatusResponse(BaseModel):
    """Response model for evidence status"""
    evidence_id: str
    status: str
    error_message: Optional[str] = None
    updated_at: Optional[str] = None


class RetryResponse(BaseModel):
    """Response model for retry endpoint"""
    success: bool
    message: str
    evidence_id: str
    status: str


router = APIRouter()


@router.post("/presigned-url", response_model=PresignedUrlResponse, status_code=status.HTTP_200_OK)
def generate_presigned_upload_url(
    request: PresignedUrlRequest,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db)
):
    """
    Generate S3 presigned URL for evidence file upload

    **Request Body:**
    - case_id: Case ID (required)
    - filename: Original filename (required)
    - content_type: MIME type (required)

    **Response:**
    - upload_url: S3 presigned POST URL
    - fields: Form fields for multipart upload
    - evidence_temp_id: Temporary evidence ID

    **Errors:**
    - 401: Not authenticated
    - 403: User does not have access to case
    - 404: Case not found

    **Security:**
    - Requires valid JWT token
    - User must be a member of the case
    - Presigned URL expires in 5 minutes max
    - Direct upload to S3 (no proxy through backend)

    **Usage:**
    Frontend should use the returned upload_url and fields to POST
    the file directly to S3 using multipart/form-data.
    """
    evidence_service = EvidenceService(db)
    return evidence_service.generate_upload_presigned_url(request, user_id)


@router.post("/upload-complete", response_model=UploadCompleteResponse, status_code=status.HTTP_201_CREATED)
def handle_upload_complete(
    request: UploadCompleteRequest,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db)
):
    """
    Notify backend that S3 upload is complete and create evidence record

    **Request Body:**
    - case_id: Case ID (required)
    - evidence_temp_id: Temporary evidence ID from presigned-url response (required)
    - s3_key: S3 object key where file was uploaded (required)
    - note: Optional note about the evidence

    **Response:**
    - evidence_id: Created evidence ID
    - case_id: Case ID
    - filename: Extracted filename
    - s3_key: S3 object key
    - status: "pending" (waiting for AI processing)
    - created_at: Creation timestamp

    **Errors:**
    - 401: Not authenticated
    - 403: User does not have access to case
    - 404: Case not found

    **Process:**
    1. Validates user has access to the case
    2. Creates evidence metadata record in DynamoDB
    3. AI Worker will automatically process when triggered by S3 event

    **Security:**
    - Requires valid JWT token
    - User must be a member of the case
    """
    evidence_service = EvidenceService(db)
    return evidence_service.handle_upload_complete(request, user_id)


@router.get("/{evidence_id}", response_model=EvidenceDetail, status_code=status.HTTP_200_OK)
def get_evidence_detail(
    evidence_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db)
):
    """
    Get detailed evidence metadata with AI analysis results

    **Path Parameters:**
    - evidence_id: Evidence ID

    **Response:**
    - Complete evidence metadata including:
      - Basic info: id, case_id, type, filename, s3_key
      - AI analysis: ai_summary, labels, insights, content (STT/OCR)
      - Timestamps: created_at, timestamp (event time in evidence)

    **Errors:**
    - 401: Not authenticated
    - 403: User does not have access to case
    - 404: Evidence not found

    **Security:**
    - Requires valid JWT token
    - User must be a member of the case that owns this evidence

    **Notes:**
    - AI analysis fields (ai_summary, labels, insights, content) are available
      only when status="done" (AI Worker processing completed)
    - For pending/processing evidence, these fields will be null
    - Full text content (STT/OCR) is included for display and manual review
    """
    evidence_service = EvidenceService(db)
    return evidence_service.get_evidence_detail(evidence_id, user_id)


@router.get("/{evidence_id}/status", response_model=EvidenceStatusResponse, status_code=status.HTTP_200_OK)
def get_evidence_status(
    evidence_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db)
):
    """
    Get current processing status of evidence

    **Path Parameters:**
    - evidence_id: Evidence ID

    **Response:**
    - evidence_id: Evidence ID
    - status: Current status (pending, uploaded, processing, completed, failed)
    - error_message: Error message if status is "failed"
    - updated_at: Last update timestamp

    **Errors:**
    - 401: Not authenticated
    - 403: User does not have access to case
    - 404: Evidence not found

    **Status Values:**
    - pending: Upload URL generated, waiting for file
    - uploaded: File uploaded, waiting for processing
    - processing: AI Worker is processing
    - completed: Processing complete
    - failed: Processing failed (can retry)
    """
    evidence_service = EvidenceService(db)
    try:
        result = evidence_service.get_evidence_status(evidence_id, user_id)
        return EvidenceStatusResponse(**result)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))


@router.post("/{evidence_id}/retry", response_model=RetryResponse, status_code=status.HTTP_200_OK)
def retry_evidence_processing(
    evidence_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db)
):
    """
    Retry processing for failed evidence

    **Path Parameters:**
    - evidence_id: Evidence ID to retry

    **Response:**
    - success: Whether retry was initiated successfully
    - message: Status message
    - evidence_id: Evidence ID
    - status: New status after retry attempt

    **Errors:**
    - 400: Evidence not in failed state (cannot retry)
    - 401: Not authenticated
    - 403: User does not have access to case
    - 404: Evidence not found

    **Process:**
    1. Validates evidence is in "failed" or "pending" state
    2. Updates status to "processing"
    3. Re-invokes AI Worker Lambda
    4. If invocation fails, status returns to "failed"

    **State Machine:**
    - FAILED → PROCESSING (on retry)
    - PENDING → PROCESSING (on retry)
    """
    evidence_service = EvidenceService(db)
    try:
        result = evidence_service.retry_processing(evidence_id, user_id)
        return RetryResponse(**result)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

File: frontend/src/lib/api/evidence.ts
/**
 * Evidence API Client
 * Handles evidence upload and retrieval operations
 */

import { apiRequest, ApiResponse } from './client';

export interface Evidence {
  id: string;
  case_id: string;
  type: 'image' | 'audio' | 'video' | 'text' | 'pdf';
  filename: string;
  s3_key: string;
  size: number;
  content_type: string;
  status: 'pending' | 'processing' | 'processed' | 'failed';
  timestamp?: string;
  speaker?: string;
  ai_summary?: string;
  labels?: string[];
  insights?: string[];
  article_840_tags?: {
    categories: string[];
    confidence: number;
    matched_keywords: string[];
  };
  created_at: string;
}

export interface EvidenceDetail extends Evidence {
  content?: string; // STT/OCR 전문 텍스트
  qdrant_id?: string;
}

export interface EvidenceListResponse {
  evidence: Evidence[];
  total: number;
}

export interface PresignedUrlResponse {
  upload_url: string;
  evidence_temp_id: string;
  s3_key: string;
  fields?: Record<string, string>;
}

export interface UploadCompleteRequest {
  case_id: string;
  evidence_temp_id: string;
  s3_key: string;
  file_size: number;
  note?: string;
}

export interface UploadCompleteResponse {
  evidence_id: string;
  case_id: string;
  filename: string;
  s3_key: string;
  status: string;
  created_at: string;
}

/**
 * Get list of evidence for a case
 */
export async function getEvidence(
  caseId: string,
  filters?: {
    categories?: string[];
    type?: Evidence['type'];
  }
): Promise<ApiResponse<EvidenceListResponse>> {
  const params = new URLSearchParams();
  if (filters?.categories) {
    params.append('categories', filters.categories.join(','));
  }
  if (filters?.type) {
    params.append('type', filters.type);
  }

  const queryString = params.toString();
  const url = `/cases/${caseId}/evidence${queryString ? `?${queryString}` : ''}`;

  return apiRequest<EvidenceListResponse>(url, {
    method: 'GET',
  });
}

/**
 * Get a single evidence item by ID (basic info)
 */
export async function getEvidenceById(
  evidenceId: string
): Promise<ApiResponse<Evidence>> {
  return apiRequest<Evidence>(`/evidence/${evidenceId}`, {
    method: 'GET',
  });
}

/**
 * Get evidence detail with full content (STT/OCR text)
 */
export async function getEvidenceDetail(
  evidenceId: string
): Promise<ApiResponse<EvidenceDetail>> {
  return apiRequest<EvidenceDetail>(`/evidence/${evidenceId}`, {
    method: 'GET',
  });
}

/**
 * Get presigned URL for uploading evidence
 */
export async function getPresignedUploadUrl(
  caseId: string,
  filename: string,
  contentType: string
): Promise<ApiResponse<PresignedUrlResponse>> {
  return apiRequest<PresignedUrlResponse>('/evidence/presigned-url', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      case_id: caseId,
      filename,
      content_type: contentType
    }),
  });
}

/**
 * Notify backend that S3 upload is complete
 */
export async function notifyUploadComplete(
  request: UploadCompleteRequest
): Promise<ApiResponse<UploadCompleteResponse>> {
  return apiRequest<UploadCompleteResponse>('/evidence/upload-complete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  });
}

export interface UploadProgress {
  loaded: number;
  total: number;
  percent: number;
}

/**
 * Upload file directly to S3 using presigned URL with progress tracking
 */
export async function uploadToS3(
  presignedUrl: string,
  file: File,
  onProgress?: (progress: UploadProgress) => void
): Promise<boolean> {
  return new Promise((resolve) => {
    const xhr = new XMLHttpRequest();

    xhr.upload.addEventListener('progress', (e) => {
      if (e.lengthComputable && onProgress) {
        onProgress({
          loaded: e.loaded,
          total: e.total,
          percent: Math.round((e.loaded / e.total) * 100),
        });
      }
    });

    xhr.addEventListener('load', () => {
      resolve(xhr.status >= 200 && xhr.status < 300);
    });

    xhr.addEventListener('error', () => {
      console.error('S3 upload error');
      resolve(false);
    });

    xhr.open('PUT', presignedUrl);
    xhr.setRequestHeader('Content-Type', file.type);
    xhr.send(file);
  });
}

/**
 * Delete evidence
 */
export async function deleteEvidence(
  evidenceId: string
): Promise<ApiResponse<void>> {
  return apiRequest<void>(`/evidence/${evidenceId}`, {
    method: 'DELETE',
  });
}

export interface RetryResponse {
  evidence_id: string;
  job_id: string;
  status: string;
  message: string;
}

/**
 * Retry processing for failed evidence
 * Re-queues the evidence for AI Worker processing
 */
export async function retryEvidence(
  evidenceId: string
): Promise<ApiResponse<RetryResponse>> {
  return apiRequest<RetryResponse>(`/evidence/${evidenceId}/retry`, {
    method: 'POST',
  });
}

export interface EvidenceStatusResponse {
  evidence_id: string;
  status: Evidence['status'];
  progress?: number;
  error_message?: string;
  retry_count?: number;
  max_retries?: number;
}

/**
 * Get the current processing status of evidence
 */
export async function getEvidenceStatus(
  evidenceId: string
): Promise<ApiResponse<EvidenceStatusResponse>> {
  return apiRequest<EvidenceStatusResponse>(`/evidence/${evidenceId}/status`, {
    method: 'GET',
  });
}
